File: src/main/java/com/heybys/oddments/base/domain/BaseEntityAware.java
package com.heybys.oddments.base.domain;

import java.time.LocalDateTime;

import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.heybys.oddments.security.OAuth2UserPrincipal;

public interface BaseEntityAware {

    @JsonIgnore
    default String getCurrentAuditor() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null && (authentication.getPrincipal()) instanceof OAuth2UserPrincipal) {
            OAuth2UserPrincipal principal = (OAuth2UserPrincipal) authentication.getPrincipal();
            return String.valueOf(principal.getId());
        }
        if (authentication != null && (authentication.getPrincipal()) instanceof JwtAuthenticationToken) {
            JwtAuthenticationToken token = (JwtAuthenticationToken) authentication.getPrincipal();
            return String.valueOf(token.getToken().getId());
        }
        return "SYSTEM";
    }

    @JsonIgnore
    default LocalDateTime getNow() {
        return LocalDateTime.now();
    }
}

File: src/main/kotlin/com/heybys/oddments/base/domain/BaseEntityAware.kt
package com.heybys.oddments.base.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import com.heybys.oddments.security.OAuth2UserPrincipal
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import java.time.LocalDateTime

interface BaseEntityAware {

    @get:JsonIgnore
    val currentAuditor: String
        get() {
            val authentication = SecurityContextHolder.getContext().authentication
            return when (val principal = authentication?.principal) {
                is OAuth2UserPrincipal -> principal.getId().toString()
                is JwtAuthenticationToken -> principal.token.id.toString()
                else -> "SYSTEM"
            }
        }

    @get:JsonIgnore
    val now: LocalDateTime
        get() = LocalDateTime.now()
}
